### PR TITLE
Support transaction import for securities with different currencies

### PR DESCRIPTION
**Pull request by inv-trad**
Issue: https://forum.portfolio-performance.info/t/buchungen-importieren-fuer-doppelt-angelegtes-wertpapier-mit-unterschiedlichen-waehrungen/13301

Im Forum kam schon mehrfach die Frage auf, warum ein Import für Dokumente mit gleichen Wertpapieren aber unterschiedlichen Währungen nicht auch funktioniert. Hierzu diese Erweiterung als Vorschlag.

Und: Der hinzugefügte Test `testThatSecurityIsMatchedByISINandName` lief erst in einen Fehler, weil die Exception for der `lookupSecurityByName` geworfen wurde. Deswegen habe ich den Lookup nach vorne gezogen.

### CODE REVIEW

**Comment by inv-trad** [comment]
Sehe gerade, dass CSVSecurityExtractorTest einen Fehler produziert. Schaue mir das gleich noch mal an.

**Comment by inv-trad** [comment]
@buchen BTW: gibt es eine Möglichkeit alle Tests einfach vorab lokal in Eclipse zu starten?

**Comment by inv-trad** [comment]
Eine offene Frage für mich ist, ob dann die PDF Importer noch so angepasst werden sollten, dass die Währung nach der gesucht wird, immer der Transaktionswährung entspricht. Den Umbau des Consorsbank Importers habe ich hinzugefügt.

File: name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/consorsbank/ConsorsbankPDFExtractorTest.java [hunk 1]
```diff
@@ -325,16 +325,12 @@ public void testErtragsgutschrift10() throws IOException
         assertThat(t.getSecurity().getName(), is("OMNICOM GROUP INC. Registered Shares DL -,15"));
         assertThat(t.getSecurity().getIsin(), is("US6819191064"));
         assertThat(t.getSecurity().getWkn(), is("871706"));
-        assertThat(t.getSecurity().getCurrencyCode(), is(CurrencyUnit.USD));
+        assertThat(t.getSecurity().getCurrencyCode(), is(CurrencyUnit.EUR));
```
Comment (buchen) [inline name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/consorsbank/ConsorsbankPDFExtractorTest.java:328]: I am not sure why we should change the currency of the created security from USD to EUR.

If you look at the extracted text, it is clear that the security is in USD:

```
OMNICOM GROUP INC. Registered Shares DL -,15 871706 US6819191064
Bestand
25 Stück
Dividende pro Stück 0,60 USD Schlusstag 17.12.2017
Brutto in USD 15,00 USD
```

**Review by buchen — COMMENTED** [review]
From my point of view, we should change only one the two things:

Yes, it can make sense to improve the matching of securities. If multiple securities are found by the same ISIN, make a decision based on the currency.

No, I think we should not change the how the security is extracted from PDF. If the PDF indicates, it is in USD and it includes exchange rate information, we should not throw it away.


Now I am thinking how to reconcile this change with #1915 by @chrisaut. The latter is proposing that the currency of dividends can be different to the currency of the security. That would mean you can have USD dividends for a EUR security (where the security is actually the currency of the prices - in which currency the security is traded. I wonder if this change "picking the security based on a security" then still makes sense.

I wonder if we should evolve PP in this direction:
* the *currency* of the security indicates only the currency of the prices - I need this because most source of historical prices do not include currency information
* enable dividend payments in arbitrary currency
* enable purchase/sales in arbitrary currency as well (one could by at different exchange)

If we have his, then the answer to the user in the forum is:
* just have *one* security
* in one securities account you trade in USD, in the other in EUR
* upon import, the dividends are attached to this one security (of course, choose different cash accounts based on the type of the document - one bank, other bank)

File: name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/consorsbank/ConsorsbankPDFExtractorTest.java (hunk 1 above)
Comment (inv-trad) [reply name.abuchen.portfolio.tests/src/name/abuchen/portfolio/datatransfer/pdf/consorsbank/ConsorsbankPDFExtractorTest.java]: I rolled back the changes to the importer and the corresponding test case.

**Comment by inv-trad** [comment]
Thank you for your comprehensive feedback. 

> I wonder if we should evolve PP in this direction:
>   -   the currency of the security indicates only the currency of the prices - I need this because most source of historical prices do not include currency information
>   -   enable dividend payments in arbitrary currency
>   -   enable purchase/sales in arbitrary currency as well (one could by at different exchange)

I agree to those points, because if you have e.g. Apple shares in a Germany portfolio (Depot) and bought it in EUR, you want to see the prices in EUR, but still it might be that - like it is done at Consorsbank - you get the dividends statement showing USD and the calculation to the amount paid in EUR to your account.

With regards to the "one-security-approach", I am not that convinced. As an example, if you have Apple shares in two different portfolios (EUR and USD) you can create two securities (EUR and USD) and have the correct prices and bookings per security and currency in the corresponding portfolio.

As I removed the change to the importer, the current proposal improves the matching 

1. to lookup by name, if e.g. ISIN has a duplicate and the name can resolve this, and 
2. to match by ISIN and currency, if ISIN has a duplicate that can be resolved by currency.

**Comment by chirlu** [comment]
> BTW: gibt es eine Möglichkeit alle Tests einfach vorab lokal in Eclipse zu starten?

Passiert bei mir automatisch. Hast du Infinitest installiert?

**Comment by chirlu** [comment]
> With regards to the "one-security-approach", I am not that convinced. As an example, if you have Apple shares in two different portfolios (EUR and USD) you can create two securities (EUR and USD) and have the correct prices and bookings per security and currency in the corresponding portfolio.

Disadvantages of this approach:
* There is no comprehensive overview of the share’s performance. It’s artificially divided into “USD shares” and “EUR shares”.
* You need to do everything twice: Add two shares to the same classifications, add twice the same links, set up historical quotes twice etc.
* It’s fully possible to buy shares on XETRA, in EUR, then sell those same shares on NYSE, in USD.
* Conversely, you may have shares bought in EUR and shares bought in USD in the same portfolio. They should not be considered separate for the Trades report.

**Comment by inv-trad** [comment]
> Passiert bei mir automatisch. Hast du Infinitest installiert?

Inifinitest hatte ich nicht installiert. Ist jetzt nachgerüstet. Danke.

**Pull request closed without merging**